--- a/gc.cc
+++ b/gc.cc
@@ -55,8 +55,8 @@

     while (1) {
         if (pathExists(fnTempRoots))
-            /* It *must* be stale, since there can be no two
-               processes with the same pid. */
+            /* The file is stale since each LocalStore instance
+               uses a unique filename (pid + counter). */
             tryUnlink(fnTempRoots);

         *fdTempRoots = openLockFile(fnTempRoots, true);
--- a/local-store.cc
+++ b/local-store.cc
@@ -113,6 +113,8 @@
     SQLiteStmt QueryPathFromHashPart;
     SQLiteStmt QueryValidPaths;
 };
+
+static std::atomic<unsigned int> nrLocalStores{0};

 LocalStore::LocalStore(ref<const Config> config)
     : Store{*config}
@@ -124,7 +126,7 @@
     , reservedPath(dbDir / "reserved")
     , schemaPath(dbDir / "schema")
     , tempRootsDir(config->stateDir.get() / "temproots")
-    , fnTempRoots(tempRootsDir / std::to_string(getpid()))
+    , fnTempRoots(tempRootsDir / fmt("%d-%d", getpid(), nrLocalStores++))
 {
     auto state(_state->lock());
     state->stmts = std::make_unique<State::Stmts>();
